lib/engine/game/g_18_al.rb

```diff
@@ -48,11 +56,66 @@ def operating_round(round_num)
         ], round_num: round_num)
       end
 
+      def stock_round
+        Round::Stock.new(self, [
+          Step::DiscardTrain,
+          Step::G18AL::BuySellParShares,
+        ])
+      end
+
       def revenue_for(route)
         # Mobile and Nashville should not be possible to pass through
         ensure_termini_not_passed_through(route, %w[A4 Q2])
 
-        adjust_revenue_for_4d_train(route, super)
+        revenue = adjust_revenue_for_4d_train(route, super)
+
+        route.corporation.abilities(:hexes_bonus) do |ability|
+          revenue += route.stops.sum { |stop| ability.hexes.include?(stop.hex.id) ? ability.amount : 0 }
+        end
+
+        route_bonuses.each do |type|
+          revenue += route_bonus(route, type)
+        end
+
+        revenue
+      end
+
+      def routes_revenue(routes)
+        # Ensure we only get a route_bonus one time, even if several routes qualify
+        total_revenue = super
+        route_bonuses.each do |type|
+          bonus_received_for_this_type = routes.sum { |r| route_bonus(r, type) }
+          next if bonus_received_for_this_type.zero?
+
+          bonus_amount = 0
+          routes.first.corporation.abilities(type).each { |a| bonus_amount = a.amount }
+          next if bonus_received_for_this_type == bonus_amount
+
+          total_revenue -= bonus_received_for_this_type
+          total_revenue += bonus_amount
```

Comment:
i'm having trouble understanding this piece of code, could it be cleaned up a bit?

why are you subtracting the bonus received for this type and then adding the bonus amount

Reply:
The idea here is that if you have two or three trains that run from Birmingham to Atlanta, you should only receive the Birmingham to Atlanta once.

So I go through all routes and add the bonuses received (which have been added to each route specifically) and if this amount is larger than the bonus one time, I need to deduct the extra amount received.

Now this is a rare occurance, but it might happen. 

Anyway, I have rewritten the code to make it clearer (hopefully)